### ISSUE TITLE
Log-cache process periodically truncates entire cache at once

### ISSUE THREAD

**Issue by crsimmons** [cloudfoundry/log-cache-release#50]
On our environments we've noticed that sometimes log-cache processes will over-aggressively truncate themselves almost completely.

<img width="1426" alt="image" src="https://user-images.githubusercontent.com/5785492/130077831-a5b79a7e-541e-4472-a13b-509ea2dc311c.png">

This happens both on dev foundations with a single doppler and minimal log load as well as on production foundations with 50+ dopplers and hundreds of thousands of logs/s. In foundations with multiple dopplers this will occur on the individual log-cache processes periodically but they won't all drop to 0% at the same time.

Sometimes it doesn't drop all the way to 0% but rather to 15 or 20%. We see this happening every day and in one of our production foundations with 60 dopplers we have seen this happen 5 times in the last 12 hours - each time on a different doppler instance.

Looking at the source code it should only be pruning a percentage of the logs as needed to stay under the configured max memory (50% by default).

When the issue occurs we don't find anything logged in any of the log-cache log files on the doppler vm. We added some extra log statements to the release in an attempt to understand what is happening. We printed the following values:

* `heapPercentage`: [the memory value used to determine how many logs to prune](https://github.com/cloudfoundry/log-cache-release/blob/main/src/internal/cache/store/prune_consultant.go#L38)
* `storeCount`: [the total number of envelopes in the cache as tracked by an atomic counter](https://github.com/cloudfoundry/log-cache-release/blob/main/src/internal/cache/store/store.go#L238)
* `numberToPrune`: [the number of envelopes to be pruned based on memory calculations](https://github.com/cloudfoundry/log-cache-release/blob/main/src/internal/cache/store/store.go#L240)
* `totalSize`: the number of envelopes in the cache as tracked by the size of the underlying AVL tree. We calculated this with the following:

  ```go
	totalSize := 0
	store.storageIndex.Range(func(sourceId interface{}, tree interface{}) bool {
		totalSize += tree.(*storage).Size()
		return true
	})
  ```

During the event graphed above our logs showed:

```log
2021/08/19 10:31:21.674552 heapPercentage: 49.971162 storeCount: 1525708
2021/08/19 10:31:21.674660 totalSize: 1525708
2021/08/19 10:31:22.175259 heapPercentage: 50.002868 storeCount: 1525712
2021/08/19 10:31:22.175365 totalSize: 1525712
2021/08/19 10:31:22.175373 numberToPrune: 87 storeCount: 1525712
2021/08/19 10:31:22.676094 heapPercentage: 50.044378 storeCount: 1525641
2021/08/19 10:31:22.677029 totalSize: 1525641
2021/08/19 10:31:22.677042 numberToPrune: 1352 storeCount: 1525641
2021/08/19 10:31:23.180395 heapPercentage: 50.079422 storeCount: 1524297
2021/08/19 10:31:23.180508 totalSize: 1524297
2021/08/19 10:31:23.180516 numberToPrune: 2417 storeCount: 1524297
2021/08/19 10:31:23.684011 heapPercentage: 50.126356 storeCount: 1521927
2021/08/19 10:31:23.684103 totalSize: 1521927
2021/08/19 10:31:23.684109 numberToPrune: 3836 storeCount: 1521927
...
# many more similar log lines with heapPercentage growing
...
2021/08/19 10:32:46.787065 heapPercentage: 56.108410 storeCount: 741
2021/08/19 10:32:46.787095 totalSize: 741
2021/08/19 10:32:46.787101 numberToPrune: 80 storeCount: 741
2021/08/19 10:32:47.287982 heapPercentage: 56.112374 storeCount: 752
2021/08/19 10:32:47.288028 totalSize: 752
2021/08/19 10:32:47.288035 numberToPrune: 81 storeCount: 752
2021/08/19 10:32:47.788980 heapPercentage: 56.114042 storeCount: 705
2021/08/19 10:32:47.789021 totalSize: 705
2021/08/19 10:32:47.789027 numberToPrune: 76 storeCount: 705
2021/08/19 10:32:48.289605 heapPercentage: 38.186652 storeCount: 732
2021/08/19 10:32:48.289659 totalSize: 732
2021/08/19 10:32:48.790157 heapPercentage: 0.262202 storeCount: 799
2021/08/19 10:32:48.790207 totalSize: 799
2021/08/19 10:32:49.291170 heapPercentage: 0.262411 storeCount: 822
2021/08/19 10:32:49.291227 totalSize: 822
```

Notice how `heapPercentage` continuously increases as the number of envelopes in the cache drops by over a million then suddenly over the course of a second it drops from 56% to 0.26%. This was on a dev foundation with one doppler vm. At the time of this activity log ingress into log-cache was averaging around 200 envelopes/s. Sometimes we see the reported heapPercentage increase above 75% before resetting to 0.

It appears that either the truncation or the change to the memory used by log-cache is delayed.

Is this expected? If so, why does it happen? If not, does anyone know what we could look at to debug this further?

**Comment by crsimmons** [comment]
We did some further research into go GC and think we might have figured it out.

We believe that when the truncation loop deletes elements from the AVL tree the allocated memory on the heap doesn't actually get freed until GC runs. This happens at least [every two minutes](https://github.com/golang/go/blob/master/src/runtime/proc.go#L5230) or more often based on some memory calculations. From our observations of log-cache the GC is triggering around once per minute.

This means that when the heap memory goes above 50% and the truncation loop starts pruning old envelopes every tick (500ms) this doesn't get reflected as a drop in allocated memory until two minutes later in the worst case. This is potentially 120 successive prunes after which almost the whole tree has been deleted and memory drops to 0 after GC.

As a test we added `runtime.GC()` to the end of the truncate function and it appears to work as expected now. When heap memory goes above 50% there is one prune operation to remove some old envelopes (which takes about a second) then heap memory drops and the truncation loop starts ticking every 500ms again.

Is there some potential problem with manually running GC after every truncation that we haven't thought of? If not I'd be happy to PR the change.

**Comment by Benjamintf1** [comment]
@crsimmons running gc loops too often could be too high of a performance cost I think?

**Comment by Benjamintf1** [comment]
Feel free to make a pr!

**Comment by crsimmons** [comment]
Running GC on ever prune does indeed cause a noticable CPU jump. I think the solution is probably somewhere in between. We were looking at two different approaches but I got pulled onto other things and never finished investigating.

1. Run GC every Nth consecutive prune. This would prevent the runaway pruning that empties the cache while hopfully limiting the CPU overhead
2. Fiddle around with `GOGC` at runtime to encourage automatic GC to trigger more often. Everything I read online suggests this could be a bad idea though.

**Comment by Benjamintf1** [comment]
Any reason why you're thinking to run the GC every nth prune rather then running it every prune but pruning less often(and configurably?)?

**Comment by crsimmons** [comment]
That's a good point. Bumping the interval would reduce the worst case number of prunes that could happen before GC kicks in. You wouldn't want to increase it too much though otherwise memory might fill completely. I'm taking another look at this and will give it a shot.

**Comment by crsimmons** [comment]
As an update I've managed to reproduce the issue in a dev environment with a single doppler (2cpu 4gb ram). By sending between 5k and 7.5k 2kb logs/s through I can regularly see the repeated truncation leading to most of the cache being emptied.

I've implemented a change making the truncation interval configurable but it doesn't appear to fix the issue. Raising the interval above 2s (default is 500ms) actually makes it worse since the VM's memory fills too fast. I think I will still PR my change since it seems like something that makes sense to be configurable. I'll continue to investigate the GC options.
